PyGA/GA.py

```diff
@@ -0,0 +1,605 @@
+        
+        return parents
+    
+    
+    ### Operators ############################################################
+    
+    def crossover(self, parent1, parent2):
+        '''Crossover two parents to create two children'''
+        antenna_type = self.settings['a_type']
+        
+        valid_children = False
+        cross_attempt = 0
+        while not valid_children:
+            child1_genes = []
+            child2_genes = []
+            
+            # Crossover genes
+            for gene in range(len(parent1.genes)):
+                gene_1 = parent1.genes[gene]
+                gene_2 = parent2.genes[gene]
+                
+                coinflip = random.randint(0, 1)
+                child1_genes.append(gene_1 if coinflip == 0 else gene_2)
+                child2_genes.append(gene_2 if coinflip == 0 else gene_1)
+                
+            child1 = self.make_antenna(antenna_type, child1_genes)
+            child2 = self.make_antenna(antenna_type, child2_genes)
+            
+            # Check if children are valid
+            valid_children = child1.check_genes() and child2.check_genes()
+            
+            cross_attempt += 1
+            if cross_attempt > 100:
+                print("Crossover Failed") if self.settings["verbose"] else None
+                return parent1, parent2
+            
+        return child1, child2
+                
+    
+    def mutation(self, individual):
+        '''Mutate a randomly selected gene across a gaussian distribution'''
+        chosen_gene_index = random.randint(0, len(individual.genes) - 1)
+        chosen_gene = individual.genes[chosen_gene_index]
+        new_indiv = copy.deepcopy(individual)
+            
+        
+        valid_antenna = False
+        while not valid_antenna:
+            new_gene = random.gauss(chosen_gene, chosen_gene * self.settings["sigma"])
+            new_indiv.genes[chosen_gene_index] = new_gene
+            valid_antenna = new_indiv.check_genes()
+            
+        return new_indiv
+    
+    
+    def reproduction(individual):
+        '''Asexual Reproduction'''
+        return individual
+    
+    
+    def injection(self):
+        '''Injects new individuals into the population'''
+
+        individual = self.make_antenna(self.settings['a_type'])
+        individual.initialize()
+        
+        return individual
+    
+    
+    ### Write/ReadFunctions ##################################################
+    
+    def write_population_genes(self, filepath=None):
+        '''Write the gened of the population to the run directory'''
+        if filepath is None:
+            filepath = self.run_dir / "Generation_Data" / f"{self.generation}_generationDNA.csv"
+        with open (filepath, "w") as file:
+            for individual in self.population:
+                for gene in individual.genes[:-1]:
+                    file.write(f"{gene},")
+                file.write(f"{individual.genes[-1]}\n")
```

Comment:
It's skipped in the loop because it has an end line after it rather than a comma. This would probably be cleaned up automatically when using the csv module, though